### ISSUE TITLE
Problems with predicted bounding boxes

### ISSUE THREAD

**Issue by Adnation** [pierluigiferrari/ssd_keras#93]
Hello there, 

I've been trying this project however I am not able to get proper bounding boxes, So here my code

```
img_height = 512  # Height of the input images
img_width = 512  # Width of the input images
img_channels = 3  # Number of color channels of the input images
# Set this to your preference (maybe `None`). The current settings
# transform the input pixel values to the interval `[-1,1]`.
intensity_mean = 127.5
# Set this to your preference (maybe `None`). The current settings
# transform the input pixel values to the interval `[-1,1]`.
intensity_range = 127.5
n_classes = 5  # Number of positive classes
# An explicit list of anchor box scaling factors. If this is passed, it will override `min_scale` and `max_scale`.
scales = [0.08, 0.16, 0.32, 0.64, 0.96]
aspect_ratios = [0.5, 1.0, 2.0]  # The list of aspect ratios for the anchor boxes
two_boxes_for_ar1 = False  # Whether or not you want to generate two anchor boxes for aspect ratio 1
steps = None  # In case you'd like to set the step sizes for the anchor box grids manually; not recommended
offsets = None  # In case you'd like to set the offsets for the anchor box grids manually; not recommended
clip_boxes = True  # Whether or not to clip the ground truth and anchor boxes to lie entirely within the image boundaries
variances = [1.0, 1.0, 1.0, 1.0]  # The list of variances by which the encoded target coordinates are scaled
coords = 'corners'  # Whether the box coordinates to be used should be in the 'centroids' or 'minmax' format, see documentation
normalize_coords = True  # Whether or not the model is supposed to use coordinates relative to the image size

K.clear_session()

model = build_model(image_size=(img_height, img_width, img_channels),
                    n_classes=n_classes,
                    mode='training',
                    l2_regularization=0.0005,
                    scales=scales,
                    aspect_ratios_global=aspect_ratios,
                    aspect_ratios_per_layer=None,
                    two_boxes_for_ar1=two_boxes_for_ar1,
                    steps=steps,
                    offsets=offsets,
                    # clip_boxes=clip_boxes,
                    variances=variances,
                    coords=coords,
                    normalize_coords=normalize_coords,
                    subtract_mean=intensity_mean,
                    divide_by_stddev=intensity_range,
                    )
```
I've dataset of around 6K image annotated in Pascal Voc format. I also have 5 positive classes. This dataset is already augemented, it has rotated, saturated and randomly cropped images in it and their annotations. 
Here is my SSD_Encoder

```
predictor_sizes = [model.get_layer('classes4').output_shape[1:3],
                   model.get_layer('classes5').output_shape[1:3],
                   model.get_layer('classes6').output_shape[1:3],
                   model.get_layer('classes7').output_shape[1:3]]

ssd_box_encoder = SSDBoxEncoder(img_height=img_height,
                                    img_width=img_width,
                                    n_classes=n_classes,
                                    predictor_sizes=predictor_sizes,
                                    scales=scales,
                                    aspect_ratios_global=aspect_ratios,
                                    two_boxes_for_ar1=two_boxes_for_ar1,
                                    steps=steps,
                                    offsets=offsets,
                                    variances=variances,
                                    pos_iou_threshold=0.5,
                                    coords=coords,
                                    normalize_coords=normalize_coords)
```

And here is my denerator 

```
train_generator = train_dataset.generate(batch_size=batch_size,
                                         shuffle=True,
                                         train=True,
                                         ssd_box_encoder=ssd_box_encoder,
                                         convert_to_3_channels=True,
                                         equalize=False,
                                         brightness=False,
                                         flip=False,
                                         translate=False,
                                         scale=False,
                                         # This one is important because the Pascal VOC images vary in size
                                         max_crop_and_resize=(img_height, img_width, 1, 3),
                                         # This one is important because the Pascal VOC images vary in size
                                         random_pad_and_resize=(img_height, img_width, 1, 3, 0.5),
                                         random_crop=False,
                                         crop=False,
                                         resize=False,
                                         gray=False,
                                         limit_boxes=True,  # While the anchor boxes are not being clipped, the ground truth boxes should be
                                         include_thresh=0.4)
```

And this is my prediction generator

```
pred_generator = pred_dataset.generate(batch_size=1,
                                         shuffle=True,
                                         train=False,
                                         ssd_box_encoder=ssd_box_encoder,
                                         convert_to_3_channels=True,
                                         equalize=False,
                                         brightness=False,
                                         flip=False,
                                         translate=False,
                                         scale=False,
                                         # This one is important because the Pascal VOC images vary in size
                                         max_crop_and_resize=(img_height, img_width, 1, 3),
                                         # This one is important because the Pascal VOC images vary in size
                                         random_pad_and_resize=(img_height, img_width, 1, 3, 0.5),
                                         random_crop=False,
                                         crop=False,
                                         resize=False,
                                         gray=False,
                                         limit_boxes=True,
                                         include_thresh=0.4)
```

Now when I am doing prediction using this model, objects are not identified in some cases and when I try to predict on image that is rotated by 180 degree all the bounding boxes have the same coordinates. Here is the sample output 

```
array([[0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0.]], dtype=float32)
```
In another approach, in which I try to predict on 1 original image it's 90 degree and 180 degree rotation and it saturated image. In such case I am getting same multiple bounding boxes with all the points being same.

Like this 

```
{'label': 'receipt',
              'xmax': 511,
              'xmin': 511,
              'ymax': 511,
              'ymin': 511},
             {'label': 'receipt',
              'xmax': 511,
              'xmin': 511,
              'ymax': 511,
              'ymin': 511}
```

So I dont know what I am doing wrong here. 

**Note:** I have also set mode to `inference` for prediction and cords to `centroids`

Thank you for reading so far and helping.

**Comment by pierluigiferrari** [comment]
Does the model make sensible predictions on non-rotated images?

**Comment by Adnation** [comment]
Nope even on original images, there no predictions most of the times or else wrong predictions. Can you tell me anything wrong in this configuration. 

Thank for quick reply

**Comment by NicoliAraujo** [comment]
I'm having these same results for training on imdb dataset for gender recognition.

**Comment by Adnation** [comment]
1. Midway weights means the model checkpoints stored while training.
2. Validation loss seems acceptable in training but when it comes to prediction it is not where near to acceptable.
3. Wrong predictions means no predictions at all or predicting same value for multiple objects with same value of xmin, ymin, xmax, ymax.

**Comment by pierluigiferrari** [comment]
Change `coords` to `centroids` and try training again to see if that makes any difference. Also, for one concrete example where the model predicts non-sense give me both the ground truth annotations for that image as they are in the respective XML file and the predicted boxes.

**Comment by Adnation** [comment]
Can I have your email address to send those images and xmls?

**Comment by pierluigiferrari** [comment]
You can post the ground truth XML and predictions for one image here. I don't necessarily need the image itself if you don't want to share it publicly, otherwise you can post that here, too.

**Comment by Adnation** [comment]
Here is the ground truth,,
```
        <object>
		<name>Classname</name>
		<pose>Unspecified</pose>
		<truncated>0</truncated>
		<difficult>0</difficult>
		<bndbox>
			<xmin>346</xmin>
			<ymin>122</ymin>
			<xmax>3297</xmax>
			<ymax>2191</ymax>
		</bndbox>
	</object>
```

And here is the predictions

```
array([[  1.        ,   0.99999917,  14.961648  ,  -5.0528884 ,
        228.50447   , 217.2788    ],
       [  0.        ,   0.        ,   0.        ,   0.        ,
          0.        ,   0.        ],
       [  0.        ,   0.        ,   0.        ,   0.        ,
          0.        ,   0.        ],
       [  0.        ,   0.        ,   0.        ,   0.        ,
          0.        ,   0.        ],
       [  0.        ,   0.        ,   0.        ,   0.        ,
          0.        ,   0.        ]], dtype=float32)

```

**Comment by pierluigiferrari** [comment]
I was most worried about what you said about all box coordinates being the same, but that is not the case here. The values for the predicted box look normal and the other zero lines are just dummy entries.

I suspect that the predictions are not being decoded correctly. Can you share a link to the Jupyter notebook you're using?

**Comment by Adnation** [comment]
So there are couple of changes, I have also trained SSD300 which seem to do some job of identification and the output is also from SSD300. One more point to add is that this boxes are before scalling them back to original size although when I scale the boxes to original they are not in place. Here is my prediction file you can look into.
```

THRESHOLD_WIDTH = 300
THRESHOLD_HEIGHT = 300
input_shape = (1, 300, 300, 3)
WEIGHTS_LOCATION = "trained_models/ssd300_weights_epoch-11_loss-1.1225_val_loss-1.3007.h5"
classes = ["class1", "class2", "class3", "class4", "class5"]

# Method to do object detection
def detect_objects(model, input_path, output_path):

    bounding_boxes = {"original" = []}

    # Read image file
    imgcv = cv2.imread(input_path)

    # Get image height, width and ignore channel
    height, width, _ = imgcv.shape
    height_factor, width_factor = 1, 1

    if height > THRESHOLD_HEIGHT:
        height_factor = height / THRESHOLD_HEIGHT
        height = THRESHOLD_HEIGHT

    if width > THRESHOLD_WIDTH:
        width_factor = width / THRESHOLD_WIDTH
        width = THRESHOLD_WIDTH

    # Scale down image for faster calculation
    imgcv_rescaled = ia.imresize_single_image(imgcv, (int(height), int(width)))

    # Reshape image
    imgcv_reshaped = np.reshape(imgcv_rescaled, input_shape)

    # Predict for original image
    original_results = model.predict(imgcv_reshaped)

    # Decode predictions
    # original_results = decode_y2(original_results,
    #                              confidence_thresh=0.7,
    #                              iou_threshold=0.6,
    #                              top_k=10,
    #                              input_coords='centroids',
    #                              normalize_coords=True,
    #                              img_height=THRESHOLD_HEIGHT,
    #                              img_width=THRESHOLD_WIDTH)

    # Get results for original images prediction
    for result in original_results[0]:

        # cv2.rectangle(imgcv_rescaled,
        #               (int(result[2]), int(result[3])),
        #               (int(result[4]), int(result[5])),
        #               (0, 255, 0), 4)

        # Draw identified objects on image
        cv2.rectangle(imgcv,
                      (int(result[2] * width_factor), int(result[3] * height_factor)),
                      (int(result[4] * width_factor), int(result[5] * height_factor)),
                      (0, 255, 0), 4)

    # cv2.namedWindow('image', cv2.WINDOW_NORMAL)
    # cv2.imshow('image', imgcv_rescaled)
    # cv2.waitKey(0)
    # cv2.destroyAllWindows()



if __name__ == "__main__":

    img_height = 300  # Height of the input images
    img_width = 300  # Width of the input images
    img_channels = 3  # Number of color channels of the input images
    subtract_mean = [123, 117, 104]  # The per-channel mean of the images in the dataset
    swap_channels = True  # The color channel order in the original SSD is BGR
    n_classes = 5  # Number of positive classes, e.g. 20 for Pascal VOC, 80 for MS COCO
    # The anchor box scaling factors used in the original SSD300 for the Pascal VOC datasets
    scales_voc = [0.1, 0.2, 0.37, 0.54, 0.71, 0.88, 1.05]
    # The anchor box scaling factors used in the original SSD300 for the MS COCO datasets
    scales_coco = [0.07, 0.15, 0.33, 0.51, 0.69, 0.87, 1.05]
    scales = scales_voc
    aspect_ratios = [[1.0, 2.0, 0.5],
                     [1.0, 2.0, 0.5, 3.0, 1.0 / 3.0],
                     [1.0, 2.0, 0.5, 3.0, 1.0 / 3.0],
                     [1.0, 2.0, 0.5, 3.0, 1.0 / 3.0],
                     [1.0, 2.0, 0.5],
                     [1.0, 2.0, 0.5]]  # The anchor box aspect ratios used in the original SSD300; the order matters
    two_boxes_for_ar1 = True
    # The space between two adjacent anchor box center points for each predictor layer.
    steps = [8, 16, 32, 64, 100, 300]
    # The offsets of the first anchor box center points from the top and left
    # borders of the image as a fraction of the step size for each predictor
    # layer.
    offsets = [0.5, 0.5, 0.5, 0.5, 0.5, 0.5]
    limit_boxes = True  # Whether or not you want to limit the anchor boxes to lie entirely within the image boundaries
    # The variances by which the encoded target coordinates are scaled as in the original implementation
    variances = [0.1, 0.1, 0.2, 0.2]
    coords = 'centroids'  # Whether the box coordinates to be used as targets for the model should be in the 'centroids', 'corners', or 'minmax' format, see documentation
    normalize_coords = True

    K.clear_session()  # Clear previous models from memory.

    model = ssd_300(image_size=(img_height, img_width, img_channels),
                    n_classes=n_classes,
                    mode='inference',
                    l2_regularization=0.0005,
                    scales=scales,
                    aspect_ratios_per_layer=aspect_ratios,
                    two_boxes_for_ar1=two_boxes_for_ar1,
                    steps=steps,
                    offsets=offsets,
                    limit_boxes=limit_boxes,
                    variances=variances,
                    coords=coords,
                    normalize_coords=normalize_coords,
                    subtract_mean=False,
                    divide_by_stddev=None,
                    swap_channels=swap_channels,
                    confidence_thresh=0.5,
                    iou_threshold=0.1,
                    nms_max_output_size=400,
                    top_k=5)

    model.load_weights(WEIGHTS_LOCATION)
    detect_objects(model, input_path, output_path)
```

**Comment by pierluigiferrari** [comment]
Ok, but these are **a lot** of changes compared to the Jupyter training notebooks, you should have told me that from the beginning. Also, what coordinate format was the model trained to predict? I'm asking because further up you had set the coordinates to 'corners' and now to 'centroids'. In case you trained to model to learn 'corners' coordinates and try to make predictions in the 'centroids' format, everything will be wrong.

See, the thing is, the Jupyter notebooks work as they are, and they work for arbitrary datasets in the Pascal VOC format. If you decide to make changes to the code in the notebooks, that's your thing, but I don't want to spend my time debugging other people's code :slightly_smiling_face: Even in a few lines of code there could be a small bug somewhere. Every line you add or change is a potential source of error.

If I were you, I would have just used the Jupyter notebook with only the absolute minimum necessary changes, which in your case is more or less just the paths to your dataset. Once you can confirm for yourself that that works, **then** you can make changes and move the whole thing to a separate python module. And even then, once again, if I were you, I would make changes at a time and after every change I would see if I broke it or if it still works as before.

**Comment by heavenao** [comment]
Hello, I can't find the class "SSDBoxEncoder”
